### PR TITLE
Integration tests for v3

### PR DESCRIPTION
**Pull request by lox**
Port integration tests from the 2-6-stable branch to v3. See #546.

### CODE REVIEW

**Comment by lox** [comment]
I'm going to merge this monster @keithpitt. It fixes a lot of bugs currently in master.

File: bootstrap/bootstrap.go [hunk 1]
```diff
@@ -29,51 +30,70 @@ type Bootstrap struct {
...
+			break
 		}
 	}
+
+	if err := b.uploadArtifacts(); err != nil {
+		b.shell.Errorf("%v", err)
+		return shell.GetExitCode(err)
+	}
+
+	// Phase errors are where something of ours broke that merits an big red error
```
Comment (keithpitt) [inline bootstrap/bootstrap.go]: (typo) `an` => `a`

File: bootstrap/bootstrap.go [hunk 2]
```diff
@@ -329,6 +351,8 @@ func (b *Bootstrap) setUp() error {
 
 // tearDown is called before the bootstrap exits, even on error
 func (b *Bootstrap) tearDown() error {
+	b.shell.Headerf("Tearing down bootstrap")
```
Comment (keithpitt) [inline bootstrap/bootstrap.go]: This might be better a `debug` comment, or a header only if debugging is on? I think 9/10 it'll be an empty header?

Comment (lox) [reply bootstrap/bootstrap.go]: Agreed. It was debugging, will remove it entirely.

**Comment by lox** [comment]
Merged in your commit from https://github.com/buildkite/agent/compare/new-bootstrap-fixes @keithpitt.

**Pull request merged**